goldmane/pkg/aggregator/aggregator.go

```diff
@@ -475,39 +452,138 @@ func (a *LogAggregator) queryFlows(req *proto.FlowListRequest) *listResponse {
+		flows, totalPages = a.defaultIndex.List(IndexFindOpts{
+			startTimeGt: req.StartTimeGte,
+			startTimeLt: req.StartTimeLt,
+			pageSize:    req.PageSize,
+			page:        req.Page,
+			filter:      req.Filter,
+		})
+
+		// Convert the flows to proto format.
+		flowsToReturn = a.flowsToResult(flows)
 	}
 
-	// Default to time-sorted flow data.
-	flows := a.defaultIndex.List(IndexFindOpts{
-		startTimeGt: req.StartTimeGte,
-		startTimeLt: req.StartTimeLt,
-		limit:       req.PageSize,
-		page:        req.PageNumber,
-		filter:      req.Filter,
-	})
+	results := make(chan *proto.FlowListResult, 100)
+	go func() {
+		defer close(results)
+		err := chanutil.WriteWithDeadline(context.Background(), results,
+			&proto.FlowListResult{
+				Meta: &proto.ListMetadata{
+					TotalPages: int64(totalPages),
+				},
+			}, time.Second*30)
+		if err != nil {
+			logrus.WithError(err).Error("Deadline exceeded while writing flow results.")
+			return
+		}
+
+		for _, flow := range flowsToReturn {
+			err := chanutil.WriteWithDeadline(context.Background(), results,
+				&proto.FlowListResult{
+					Value: flow,
+				}, time.Second*30)
+
+			if err != nil {
+				logrus.WithError(err).Error("Deadline exceeded while writing flow results.")
+				return
+			}
+		}
+	}()
+
+	return &listResponse{results, nil}
+}
+
+func (a *LogAggregator) queryFilterHints(req *proto.FilterHintsRequest) *filterHintsResponse {
+	logrus.WithFields(logrus.Fields{"req": req}).Debug("Received flow request")
+
+	// Sanitize the time range, resolving any relative time values.
+	req.StartTimeGte, req.StartTimeLt = a.normalizeTimeRange(req.StartTimeGte, req.StartTimeLt)
 
-	// Convert the flows to proto format.
-	flowsToReturn := a.flowsToResult(flows)
-	return &listResponse{flowsToReturn, nil}
+	// Validate the request.
+	if err := a.validateTimeRange(req.StartTimeGte, req.StartTimeLt); err != nil {
+		return &filterHintsResponse{nil, err}
+	}
+
+	var sortBy proto.SortBy
+	switch req.Type {
+	case proto.FilterType_FilterTypeDestName:
+		sortBy = proto.SortBy_DestName
+	case proto.FilterType_FilterTypeDestNamespace:
+		sortBy = proto.SortBy_DestNamespace
+	case proto.FilterType_FilterTypeSourceName:
+		sortBy = proto.SortBy_SourceName
+	case proto.FilterType_FilterTypeSourceNamespace:
+		sortBy = proto.SortBy_SourceNamespace
+	default:
+		return &filterHintsResponse{nil, fmt.Errorf("unsupported filter type '%s'", req.Type.String())}
+	}
+
+	// If a sort order was requested, use the corresponding index to find the matching flows.
+	if idx, ok := a.indices[sortBy]; ok {
+		// If a sort order was requested, use the corresponding index to find the matching flows.
+		// We need to convert the FlowKey to a string for the index lookup.
+		keys, totalPages := idx.Keys(IndexFindOpts{
+			startTimeGt: req.StartTimeGte,
+			startTimeLt: req.StartTimeLt,
+			pageSize:    req.PageSize,
+			page:        req.Page,
+			filter:      req.Filter,
+		})
+
+		results := make(chan *proto.FilterHintsResult, 100)
```

Comment:
Worth doing this based on the number of results?

Like, if we know there are only 5 results we can do a smaller channel. If we know there are 1k we can do a larger channel (maybe with a hard cap at 1k or something)